File: 380.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <map>
#include <queue>
#include <random>
#include <set>
#include <string>
#include <vector>
using namespace std;

class RandomizedSet {
private:
  vector<int> v;
  map<int, int> m;
  random_device rd;
  mt19937 generator;

public:
  RandomizedSet() : generator(rd()) {}

  bool insert(int val) {
    if (m.find(val) == m.end()) {
      v.push_back(val);
      m[val] = v.size() - 1;
      return true;
    }
    return false;
  }

  bool remove(int val) {
    if (m.find(val) == m.end()) {
      return false;
    }
    if (m[val] == v.size() - 1) {
      v.pop_back();
    } else {
      int idxOfOriginal = m[val];
      int idxOfLast = v.size() - 1;
      int last = v[v.size() - 1];
      v[idxOfOriginal] = v[idxOfLast];
      m[last] = idxOfOriginal;
      v.pop_back();
    }
    m.erase(val);
    return true;
  }

  int getRandom() {
    uniform_int_distribution<int> distribution(0, v.size() - 1);
    return v[distribution(generator)];
  }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

// test 1:

int main() {
  RandomizedSet rs;
  assert(rs.remove(0) == false);
  assert(rs.remove(0) == false);
  assert(rs.insert(0) == true);
  assert(rs.remove(0) == true);
  assert(rs.insert(0) == true);
}

File: 380.py
from random import choice


class RandomizedSet:
    def __init__(self):
        self.v = []
        self.m = {}

    def insert(self, val: int) -> bool:
        if val in self.m:
            return False
        self.v.append(val)
        self.m[val] = len(self.v) - 1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.m:
            return False
        # self.v [ idx of val] = self.v[idx of last element]
        # self.m [last element] = idx of val
        # self.v.pop()
        self.v[self.m[val]] = self.v[len(self.v) - 1]
        self.m[self.v[len(self.v) - 1]] = self.m[val]
        self.v.pop()
        del self.m[val]
        return True

    def getRandom(self) -> int:
        return choice(self.v)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

# ["RandomizedSet","insert","insert","remove","insert","remove","getRandom"]
# [[],[0],[1],[0],[2],[1],[]]

obj = RandomizedSet()
obj.insert(0)
obj.insert(1)
obj.remove(0)
obj.insert(2)
obj.remove(1)
assert obj.getRandom() == 2
